src/main/scala/com/fulcrumgenomics/bam/ErrorRateByReadPosition.scala

```diff
@@ -0,0 +1,280 @@
+  * They should then call [incrementIfValid] or [increment] to update the counters for each read offset.
+  *
+  * @param minBaseQuality the minimum base quality to allow.
+  */
+private abstract class ErrorRateByReadPositionCounter(includeInsertions: Boolean, minBaseQuality: Int = 0) {
+  private val errorsByReadOffset = List.tabulate(3)(x => new NumericCounter[Int]())
+  private val totalByReadOffset  = List.tabulate(3)(x => new NumericCounter[Int]())
+
+  /** Gets the list of metrics sorted by read number and then position. */
+  def metrics: Seq[ErrorRateByReadPositionMetric] = {
+    if (totalByReadOffset.forall(_.isEmpty)) return Seq.empty
+    // create the metrics
+    errorsByReadOffset.zip(totalByReadOffset).zipWithIndex.filter {
+      case ((errCounter, totalCounter), idx) => totalCounter.nonEmpty
+    }.flatMap { case ((errCounter, totalCounter), idx) =>
+      val masOffset = totalCounter.map(_._1).max
+      Range.inclusive(0, masOffset).map { offset =>
+        new ErrorRateByReadPositionMetric(
+          read_number = idx,
+          position    = offset + 1,
+          error_rate  = errCounter.countOf(offset) / totalCounter.countOf(offset).toDouble,
+          bases_total = totalCounter.countOf(offset)
+        )
+      }
+    }.sortBy(m => (m.read_number, m.position))
+  }
+
+  /** Checks that the given bases are valid, and increments the # of errors and total bases. The read index is
+    * 0 for fragment reads, 1 for the first of a pair, and 2 for second of a pair.  The read position is zero-based. */
+  protected def incrementIfValid(readIndex: Int, readOffset: Int, readBase: Byte, qual: Byte, refBase: Byte): Unit = {
+    // Assert the bases must be valid
+    if (validBase(refBase) && validBase(readBase) && minBaseQuality <= qual) {
+      increment(readIndex=readIndex, readOffset=readOffset, qual=qual, isError = !SequenceUtil.basesEqual(readBase, refBase))
+    }
+  }
+
+  /** Increments the error count if `isError` is true. Also increments the total bases observed. The read index is
+    * 0 for fragment reads, 1 for the first of a pair, and 2 for second of a pair.  The read offset is zero-based. */
+  protected def increment(readIndex: Int, readOffset: Int, qual: Byte, isError: Boolean): Unit = {
+    if (minBaseQuality <= qual) {
+      if (isError) {
+        errorsByReadOffset(readIndex).count(readOffset)
+      }
+      totalByReadOffset(readIndex).count(readOffset)
+    }
+  }
+
+  /** Implement me! */
+  protected def updateErrors(rec: SAMRecord, readIndex: Int): Unit
+
+  def accept(rec: SAMRecord): Unit = {
+    val readIndex: Int = if (!rec.getReadPairedFlag) 0 else if (rec.getFirstOfPairFlag) 1 else 2
+    updateErrors(rec=rec, readIndex=readIndex)
+  }
+
+  protected def maybeReverseComplement(bases: Array[Byte], rc: Boolean): Array[Byte] = {
+    if (rc) SequenceUtil.reverseComplement(bases)
```

Comment:
Since the following use of it doesn't need a clone, and thus would have an _extra_ clone:
```
maybeReverseComplement(SequenceUtil.makeReferenceFromAlignment(rec, false), rc=rec.getReadNegativeStrandFlag)
```